File: lib/src/main/java/net/imoya/android/preference/model/TimePeriod.java
package net.imoya.android.preference.model;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.Locale;

/**
 * 時:分によって指定される両端を持つ期間モデル
 */
public class TimePeriod implements Parcelable {
    /**
     * 開始時
     */
    private int startHour = 0;
    /**
     * 開始分
     */
    private int startMinute = 0;
    /**
     * 終了時
     */
    private int endHour = 0;
    /**
     * 終了分
     */
    private int endMinute = 0;

    /**
     * 文字列を解析して、対応する {@link TimePeriod} を返します。
     * 文字列の書式は
     * &quot;&lt;StartHour&gt;:&lt;StartMinute&gt;-&lt;EndHour&gt;:&lt;EndMinute&gt;&quot; です。<ul>
     * <li>各項目の省略はできません。</li>
     * <li>Hourは0～23、Minuteは0～59の整数値を、10進数で表記する必要があります。</li>
     * <li>StartとEndの大小関係はチェックされません。</li>
     * </ul>
     *
     * @param s 文字列
     * @return {@link TimePeriod}
     * @throws NumberFormatException    引数 s が null です
     * @throws IllegalArgumentException 引数 s の書式が不正です。
     */
    public static TimePeriod parse(String s) throws NullPointerException, IllegalArgumentException {
        if (s == null) {
            throw new NullPointerException("s == null");
        }
        // ハイフンの位置を探す
        final int separatorPos = s.indexOf('-');
        if (separatorPos == -1) {
            throw new IllegalArgumentException("'-' not found");
        } else if (separatorPos == 0) {
            throw new IllegalArgumentException("Start time not found");
        } else if (separatorPos == s.length() - 1) {
            throw new IllegalArgumentException("End time not found");
        }
        // ハイフンで区切る
        final String start = s.substring(0, separatorPos);
        final String end = s.substring(separatorPos + 1);

        // それぞれの時:分を取得し、フィールドへ保存する
        final Time startTime = Time.parse(start);
        final Time endTime = Time.parse(end);
        final TimePeriod result = new TimePeriod();
        result.startHour = startTime.getHour();
        result.startMinute = startTime.getMinute();
        result.endHour = endTime.getHour();
        result.endMinute = endTime.getMinute();
        return result;
    }

    public TimePeriod() {
    }

//    public TimePeriod(TimePeriod source) {
//        this.copyFrom(source);
//    }

    private TimePeriod(Parcel parcel) {
        this.startHour = parcel.readByte();
        this.startMinute = parcel.readByte();
        this.endHour = parcel.readByte();
        this.endMinute = parcel.readByte();
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeByte((byte) this.startHour);
        dest.writeByte((byte) this.startMinute);
        dest.writeByte((byte) this.endHour);
        dest.writeByte((byte) this.endMinute);
    }

    public void copyFrom(TimePeriod source) {
        if (source == null)
            throw new IllegalArgumentException("source == null");
        this.startHour = source.startHour;
        this.startMinute = source.startMinute;
        this.endHour = source.endHour;
        this.endMinute = source.endMinute;
    }

    public int getStartHour() {
        return this.startHour;
    }

    public void setStartHour(int startHour) {
        if (startHour < 0 || startHour > 23) {
            throw new IllegalArgumentException("startHour requires 0 - 23");
        }
        this.startHour = startHour;
    }

    public int getStartMinute() {
        return this.startMinute;
    }

    public void setStartMinute(int startMinute) {
        if (startMinute < 0 || startMinute > 59) {
            throw new IllegalArgumentException("startMinute requires 0 - 59");
        }
        this.startMinute = startMinute;
    }

    public int getEndHour() {
        return this.endHour;
    }

    public void setEndHour(int endHour) {
        if (endHour < 0 || endHour > 23) {
            throw new IllegalArgumentException("endHour requires 0 - 23");
        }
        this.endHour = endHour;
    }

    public int getEndMinute() {
        return this.endMinute;
    }

    public void setEndMinute(int endMinute) {
        if (endMinute < 0 || endMinute > 59) {
            throw new IllegalArgumentException("endMinute requires 0 - 59");
        }
        this.endMinute = endMinute;
    }

    /**
     * 指定の時刻が、範囲内であるか否かを返します。
     *
     * @param hourOfDay 時(0～23)、又は未定義を表す-1
     * @param minute    分(0～59)、又は未定義を表す-1
     * @return 指定の時刻が範囲内の場合はtrue, 範囲外であるか未定義である場合はfalse
     */
    public boolean isInPeriod(int hourOfDay, int minute) {
        if (hourOfDay > 23 || minute > 59) {
            throw new IllegalArgumentException("Illegal time(" + hourOfDay + ":" + minute + ")");
        }
        if (this.startHour < 0 || this.startMinute < 0
                || this.endHour < 0 || this.endMinute < 0) {
            // 未定義の場合は、常に範囲外とする。
            return false;
        }

        final int target = hourOfDay * 60 + minute;
        final int from = this.startHour * 60 + this.startMinute;
        final int to = this.endHour * 60 + this.endMinute;

        // from <= to ならばその間がこのインスタンスが表す範囲である。
        // from > to ならばその間以外がこのインスタンスが表す範囲である。
        if (from <= to) {
            if (from <= target) {
                return target <= to;
            }
        } else {
            if (from <= target) {
                return true;
            }
            return target <= to;
        }

        // 以上に該当しなかったら、このインスタンスが表す範囲の中ではない。
        return false;
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof TimePeriod)) {
            return false;
        }
        final TimePeriod another = (TimePeriod) obj;
        return this.startHour == another.startHour
                && this.startMinute == another.startMinute
                && this.endHour == another.endHour
                && this.endMinute == another.endMinute;
    }

    /**
     * オブジェクトの文字列形式を得ます。
     * {@link #parse(String)} メソッドが解析可能な書式を出力します。
     *
     * @return 文字列形式
     */
    @Override
    public String toString() {
        return String.format(Locale.US, "%d:%02d-%d:%02d",
                this.startHour, this.startMinute, this.endHour, this.endMinute);
    }

    @Override
    public int hashCode() {
        return this.startHour * 86400 + this.startMinute * 1440
                + this.endHour * 60 + this.endMinute;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    /**
     * {@link Parcelable} 対応用 {@link Creator}
     */
    public static final Creator<TimePeriod> CREATOR = new Creator<TimePeriod>() {
        /**
         * {@link Parcel} の内容を保持するオブジェクトを生成して返します。
         *
         * @param parcel {@link Parcel}
         * @return {@link Parcel} の内容を保持するオブジェクト
         */
        @Override
        public TimePeriod createFromParcel(Parcel parcel) {
            return new TimePeriod(parcel);
        }

        /**
         * オブジェクトの配列を生成して返します。
         *
         * @param size 配列のサイズ
         * @return 配列
         */
        @Override
        public TimePeriod[] newArray(int size) {
            return new TimePeriod[size];
        }
    };
}

File: lib/src/main/java/net/imoya/android/preference/model/TimePeriod.kt
/*
 * Copyright (C) 2022 IceImo-P
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.imoya.android.preference.model

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator

/**
 * 時:分によって指定される両端を持つ期間モデル
 */
data class TimePeriod(
    /**
     * 開始時刻
     */
    var start: Time = Time(),
    /**
     * 終了時刻
     */
    var end: Time = Time()
) : Parcelable {
    /**
     * 値が有効であるか否かを返します。
     *
     * 値が有効であるとは [start], [end] いずれも有効である状態です。
     *
     * [Time] が有効である定義は [Time.isValid] を参照してください。
     *
     * @return 値が有効である場合は true, その他の場合は false
     * @see [Time.isValid]
     */
    @Suppress("unused")
    fun isValid(): Boolean = start.isValid() && end.isValid()

    /**
     * 指定の時刻が、範囲内であるか否かを返します。
     *
     * @param time 時刻
     * @return 指定の時刻が範囲内の場合は true, 範囲外である場合は false
     * @throws IllegalArgumentException [time] が有効ではありません([Time.isValid] が false を返しました)
     * @throws IllegalStateException この [TimePeriod] が有効ではありません([isValid] が false を返しました)
     */
    @Suppress("MemberVisibilityCanBePrivate")
    fun isInPeriod(time: Time): Boolean {
        require(time.isValid()) { "Illegal time($time)" }
        if (!isValid()) throw IllegalStateException("invalid")

        // start == end ならばその1点がこのインスタンスが表す範囲である。
        // start < end ならばその間(start, endを含む)がこのインスタンスが表す範囲である。
        // start > end ならばその間以外(start, endを含む)がこのインスタンスが表す範囲である。
        return when {
            start == end -> time == end
            start < end -> time in start..end
            else /* equivalent to start > end */ -> time >= start || time <= end
        }
    }

    /**
     * 指定の時刻が、範囲内であるか否かを返します。
     *
     * @param hourOfDay 時(0～23)
     * @param minute    分(0～59)
     * @param second    秒(0～59)
     * @return 指定の時刻が範囲内の場合は true, 範囲外である場合は false
     */
    @Suppress("unused")
    fun isInPeriod(hourOfDay: Int, minute: Int, second: Int = 0): Boolean {
        return isInPeriod(Time(hourOfDay, minute, second))
    }

    /**
     * 他の [TimePeriod] インスタンスの値のコピーをこのインスタンスに設定します。
     *
     * @param other 他のインスタンス
     */
    fun copyFrom(other: TimePeriod) {
        this.start = other.start.copy()
        this.end = other.end.copy()
    }

    override fun toString() = "$start-$end"

    override fun writeToParcel(dest: Parcel, flags: Int) {
        start.writeToParcel(dest, flags)
        end.writeToParcel(dest, flags)
    }

    override fun describeContents() = 0

    companion object {
//        /**
//         * Tag for log
//         */
//        private const val TAG = "TimePeriod"

        /**
         * 文字列を解析して、対応する [TimePeriod] を返します。
         *
         * 文字列の書式は
         * &quot;&lt;Start&gt;-&lt;End&gt;&quot;
         * (例: 0:00:00-1:23:45, 3:4-5:6, 23:59-0:01)です。
         *
         *  * Start, End の省略はできません。
         *  * Start, End は [Time] の文字列形式です。書式は [Time.parse] のドキュメントをご参照ください。
         *  * Start と End の大小関係はチェックされません。
         *  Start, End が有効であれば([Time.isValid] が true を返せば)どのような大小関係も有効です。
         *
         * @param s 文字列
         * @return [TimePeriod]
         * @throws IllegalArgumentException 引数 s の書式が不正です。
         * @see [Time.parse]
         */
        @JvmStatic
        @Throws(IllegalArgumentException::class)
        fun parse(s: String): TimePeriod {
            // ハイフンの位置を探す
            val separatorPos = s.indexOf('-')
            require(separatorPos != -1) { "'-' not found" }
            require(separatorPos != 0) { "Start time not found" }
            require(separatorPos < s.length - 1) { "End time not found" }
            // ハイフンで区切る
            val start = s.substring(0, separatorPos)
            val end = s.substring(separatorPos + 1)

            // それぞれの時:分(:秒)を取得し、フィールドへ保存する
            return TimePeriod(Time.parse(start), Time.parse(end))
        }

        /**
         * [Parcelable] 対応用 [Creator]
         */
        @JvmField
        val CREATOR: Creator<TimePeriod> = object : Creator<TimePeriod> {
            /**
             * [Parcel] の内容を保持するオブジェクトを生成して返します。
             *
             * @param parcel [Parcel]
             * @return [Parcel] の内容を保持するオブジェクト
             */
            override fun createFromParcel(parcel: Parcel): TimePeriod =
                TimePeriod(
                    Time.CREATOR.createFromParcel(parcel),
                    Time.CREATOR.createFromParcel(parcel)
                )

            /**
             * オブジェクトの配列を生成して返します。
             *
             * @param size 配列のサイズ
             * @return 配列
             */
            override fun newArray(size: Int): Array<TimePeriod?> {
                return arrayOfNulls(size)
            }
        }
    }
}